### PR TITLE
Add tensor step and capturable support to rprop

### PR DESCRIPTION
**Pull request by mlazos**
Towards fixing https://github.com/pytorch/pytorch/issues/115679
Fixes Rprop step update while compiling

Also adds capturable support + testing

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/optim/rprop.py [hunk 1]
```diff
@@ -284,8 +332,23 @@ def _multi_tensor_rprop(
 
     assert not differentiable, "_foreach ops don't support autograd"
 
-    grouped_tensors = Optimizer._group_tensors_by_device_and_dtype([params, grads, prevs, step_sizes])
-    for ((grouped_params, grouped_grads, grouped_prevs, grouped_step_sizes), _) in grouped_tensors.values():
+    # If compiling, the compiler will handle cudagraph checks, see note [torch.compile x capturable]
+    if not torch._utils.is_compiling() and capturable:
+        assert all((p.is_cuda and step.is_cuda) or (p.is_xla and step.is_xla) for p, step in zip(params, state_steps)), \
+            "If capturable=True, params and state_steps must be CUDA tensors."
```
Comment (janeyx99) [inline torch/optim/rprop.py:338]:
```suggestion
            "If capturable=True, params and state_steps must be CUDA or XLA tensors."
```

File: torch/optim/rprop.py [hunk 2]
```diff
@@ -247,17 +284,26 @@ def _single_tensor_rprop(
...
         else:
             sign = grad.mul(prev).sign()
-        sign[sign.gt(0)] = etaplus
-        sign[sign.lt(0)] = etaminus
-        sign[sign.eq(0)] = 1
+
+        if capturable:
+            sign.copy_(torch.where(sign.gt(0), etaplus, sign))
+            sign.copy_(torch.where(sign.lt(0), etaminus, sign))
+            sign.copy_(torch.where(sign.eq(0), 1, sign))
```
Comment (janeyx99) [inline torch/optim/rprop.py:291]: Not expected to be in the scope of the PR but I wonder if there's a fast way to assign all of these. Previously the index select would be pretty fast and only write into N total elements for all three lines. Now, we write into 3N total elements, in addition to calling `torch.where`. Would be interesting to have an API that would call the same lambda on every element in the Tensor to copy into, like a:
```
sign.copy_(lambda x: etaplus if x > 0 else (etaminus if x < 0 else 1))
```
This API wouldn't be general enough to replace the grad.copy_ line later though :/

**Review by janeyx99 — APPROVED** [review]
Looks good. One comment on the error message and another open-ended one I don't anticipate you'll need to change anything for.

File: torch/optim/rprop.py (hunk 2 above)
Comment (mlazos) [reply torch/optim/rprop.py:291]: if we had https://github.com/pytorch/pytorch/issues/117884 we could at least do this in two kernels which would be nice. We could possibly implement what you're talking about with a higher order op. Luckily though for compiling this doesn't appear to affect the kernel count, we still generate an optimized kernel

**Pull request closed without merging**